clang/tools/driver/cc1depscanProtocol.cpp

```diff
@@ -22,6 +23,18 @@ using namespace clang;
 using namespace clang::cc1depscand;
 using namespace llvm;
 
+namespace {
+/// Provides additional storage for a larger path name.
+struct FatSockAddr {
+  sockaddr_un Base;
+  char PathExtra[0xFF - sizeof(sockaddr_un)];
```

Comment:
> Is this trick portable?
I think so, but I'm not 100% sure.

> does this happen early during the process
Maybe that'd work; certainly can do it in the server, and I *think* it's safe in the client too (even if we're in the same process as the driver, there shouldn't really be concurrent threads). We could check if `length >= sizeof(addr.sun_path)`, and if so, chdir over and back. Man pages suggest that `connect` and `bind` take relative addresses.

Reply:
I wrote a small test using the "extended string" that works on macOS, but fails on Linux: [godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIAOykrgAyeAyYAHI%2BAEaYxCBmAJykAA6oCoRODB7evgGp6ZkCoeFRLLHxSbaY9o4CQgRMxAQ5Pn6BdpgOWQ1NBCWRMXEJyQqNza15HeP9YYPlw0kAlLaoXsTI7BzmAMxhyN5YANQmO25xxAyoAHQIp9gmGgCCu/uHmCdnY/g3dzsPz1eDAOXmOpzcY2IYWAt3ujxeZj2wPenwhAE8FAB6BSoZAAa0wBFh/3hQJBYK%2BGMxBDRKUwCmJAIRSPJH3BCipXgYjNJiLeoLZZy5eG%2BPMBz0hXgcRwAYkxhLi8U90OhiCd/FZnkdtTj8UwVcQAPpco4WJhKU6ap7ao7IBBNI7KeUIbCqAjEJgmACsFg0qhlMqOAFojhkAF6YVBUCC6vH61XGhhLb0AEUtpK12sx2aOAFkmKojilnaG8BHSEd%2BfgjEcGF5aLQjgQ4iwwvKSNd4TaxvK8MhbQIxphVCk1eHMIaCHmC06CAghGXBWnMzadYuoxBZy63R6lidLKWIxvY/GjVyQCAFFzDcW58mdlaTP40w%2BM08AG6oPDoUOKwmT%2BkCAgNAGDGW17TVAAqW8ED3J8rW1JgvCIItnUNegGE%2BFNQ3dDCIBg%2B8EKLKFBGjcwzGbMZoSOAB3QgEFQucjgw4AmNObDzC9bxvTcBhyIrGD0MYQi3yzTEjjcYhMHlD5T3pBQu1leUhEVZVVSOU900zU9rjNJRrivBhDSoVg6DRLCjieGVDTkCIAEkAA0tOtSsqCOfC0Iwz4Hh2bC5QVfE1PiEB81ULcFwjOCNUU7VRzCAgyLMMxtQuEgQEYhAhMwiBOO8OCSV8pSAqVA0L1C8LF3cziwy8ZMvV4/iMqyit/JUwLSpCmdnQizARJXI4pIIdY%2BNfTMn2XFzIWBFI0QgHS9MwAzrxggTnRa5TVI68rusXPqXkzHNJOk5tf3xQlFKQlDY0JCyL2uoCrJs%2ByHIrIQAHk3AAaUNIQABUACVsCeXMKw0PbtTwNyYz/KdwSOMH1SI2KSISnKkpS4hiDSo4TLoTAfxQ5ApJk06CVhswvXQdLOIU%2Bq%2BKSisLiuCtIVS4gICZ1AlnBm1BuG5ztXG0SjhzCwwgJwDFMh9zonF6GzoICt5bjA0jkg5MzAANlPV7bMNYJsAiNHtYNXTzV67nEZi4j4sS5KjjZ9LcfoAnUCOWWGFd0mbs4qn9y9WmGoZh3MeZnCpMxkgOdDrmeaF8UXMO%2BhDC8FJFIOdJMGVwkea5WgwjxDy72cxS4tItH7avZBNgUBQwDAHj6bMPb4/2p54qOFgmDCCAoqIkCwLtB1IIUBASAIG8S3Y/czGpFgUmpBARUnudDWXrgNB2Q1ojRSj0IEYBDQc4%2BT9Ps/z4vy%2Br%2Bvm/T/IgWB1Aqch6g2gD5Xhjp/IueF7nZfBPXmYDQ29d70n3kYI%2Bt8oHQJgbA2B99RqTRhgBMYMYx7NA/jze6KCgJvwgQRB%2BfNLjwxLs%2BeEHAVi0E4F6XgfgOBaFIKgTgbhrDWF/OsTY%2B4dg8FIAQTQFCVh4hADsHY1wRHiIkZIzW%2BhOCSFofwxhnBeAKBABoXh/CVhwFgEgNA888ZkAoMBVAej6DxAOIYYAUgNBqJoLQZsxAVEQGiAoj2TQ0ScB4botggg3oMFoO4%2BhvAsBdyMOIQJpB8BSW6O%2BekCjhxdGQlsBh8UagKPztED0xA0QeCwAo90eAWAeIoXwAwwAFAADU8CYBom9OkdCeH8EECIMQ7ApAyEEIoFQ6hwm6C4PoCxKBWGWH0HgaIKjIArFQCkOooFOBBm%2BOxUwlhrBAOUTULoMyXCe0mH4PpIQ5hlAqHoNIGQZk7OOYUGZAxDnDD6Z0bo9QZjnLuesh5DBejNGuUMeIdynmeDaHoHsnyDnfIkCsHEHDWm8KklsHglDqHyPCUwjgqgAAcmsgya0kLaAwNYpDXA0AS9yLDlnDKOLgQgJAuF9IksYlI%2BjqVLF4HwwJ3NSBCMkF6a4msvRWKStYxISVEia2kVQjgcjSB0IYci5Rqj1GstIFoxAIA1gEBSMhcglBdH0tMXofGlKfntOEKIcQbTGnyCUGoPJMKimkBoh6FIRT4UcBoZKhRyK3rIXVVOKMRw0UYqxTiixRx8WEo0MSulDLdhcCZfKrQbKhFmC4ASrgXotaSEkDy/wkgdia2SGKiVUreAytsHKll8bnVmERdKpRcaBGkBiQ4rIIBJBAA%3D%3D)

Reply:
> Maybe that'd work; certainly can do it in the server, and I _think_ it's safe in the client too (even if we're in the same process as the driver, there shouldn't really be concurrent threads)

I'd prefer that we avoid making assumptions about what the client is doing, e.g. the client itself maybe be part of a long-running process instead of a one-off driver invocation, in which case this kind of setup will come back to bite us.

Reply:
if I go back to previous patch but drop ".socket" to just ".s", I should be able to use 8 byte hash and be under the limit. Optionally I could also incorporate the pid number into the hash for more wiggle room.

Reply:
> Linux:

The frustrating thing is that on Linux:
- `TMPDIR` is going to be way shorter
- `sun_path` has a length of 108 instead of 104
- You don't even need a path on disk; you can make the address "abstract" and do `"\0llvm.depscan.whatever"` (need to compute length before setting the `\0`).

But if it doesn't work on Linux, maybe it won't work elsewhere too, and I think only Linux has abstract addresses.

> I'd prefer that we avoid making assumptions about what the client is doing, e.g. the client itself maybe be part of a long-running process instead of a one-off driver invocation, in which case this kind of setup will come back to bite us.

That's a good point. It would restrict future flexibility.

> if I go back to previous patch but drop ".socket" to just ".s", I should be able to use 8 byte hash and be under the limit. Optionally I could also incorporate the pid number into the hash for more wiggle room.

I don't love `.s` since that looks like an assembly file but `.sock` wins you a couple of characters. I think you should pull in the PID.

You could just make it `llvm-<HASH>.sock` and pull "depscan" into the hash as well. Although maybe the `depscan` is helpful for autocompletion when debugging problems.

Reply:
> But if it doesn't work on Linux, maybe it won't work elsewhere too, and I think only Linux has abstract addresses.

Interesting, I didn't know it could be abstract on Linux.  You could take a platform-specific approach instead of trying to make a single implementation work.  E.g.
* Linux: use abstract address
* macOS: use the extra long path hack
* FreeBSD: there is connectat and bindat that use an fd for the directory to look in. This would be wonderful to have on all platforms...
* Unknown posix platform: must fit in sizeof(sun_path)